content/updates/understanding-the-terraform-check-block-feature.md

```diff
@@ -0,0 +1,207 @@
+2. **Out-of-band change detection:** an unanticipated change made outside Terraform can cause surprising failures, especially if it can't be detected via a built-in drift detection mechanism (for instance, when another root module manages a resource). Checks come in handy to detect such deviations as the expiration of a TLS certificate or authentication tokens, preventing unforeseen interruptions.
+3. **Enforcing naming and tagging conventions:** maintaining a consistent set of tagging and naming conventions within your organization can be crucial. Use the `check` block feature to enforce these standards across your infrastructure, fostering uniformity and orderliness.
+4. **Upholding security best practices:** checks can help ensure adherence to security best practices within your infrastructure. For example, you can use it to verify that specific ports are closed, IAM configurations are properly set, and encryption is enabled, contributing to a more secure and robust environment.
+
+## Do I need this if I'm already testing my Terraform code?
+
+While similar in their goals of ensuring correct and efficient infrastructure operation, infrastructure validation and testing differ slightly in their scopes and methods.
+
+_Infrastructure validation_ generally refers to checking the correctness of infrastructure code against specified standards or requirements before it's deployed. For instance, native Terraform commands like `validate` or `plan` or tools like [TFLint](https://github.com/terraform-linters/tflint) mainly focus on the correctness of code and check the syntax and internal consistency, ensuring it's well-structured and free from apparent errors before deployment.
+
+_Infrastructure testing_, on the other hand, usually takes place after the infrastructure has been provisioned and validates that it's functioning as expected. This can include unit tests, integration tests, functional tests, and acceptance tests, among others. Testing tools like [Kitchen-Terraform](https://newcontext-oss.github.io/kitchen-terraform/) and [Terratest](https://terratest.gruntwork.io/) are commonly used for these kinds of tests. They can confirm that resources are adequately created and connected, security groups allow/deny correct traffic, etc.
+
+Most infrastructure tools use declarative language to define the desired state. By its nature, declarative code does not specify the steps needed to reach this state. Inspecting the desired state's details to confirm proper changes while writing the tests can quickly become tedious.
+
+This is where `check` blocks come into play, stepping in to handle assertions when the actual provisioning isn't the primary focus. Given the necessity of configuring external tools and writing tests in Ruby or Go, native HCL `check` blocks can be seen as a practical alternative. They can be used before running tests in sandbox environments — whether ephemeral or persistent, hosted in the cloud, or run locally using mocking tools.
```

Comment:
This paragraph is a bit confusing to me and it is a critical paragraph. Do you mind explaining a couple things and we can hopefully clarify it a bit? 

> In fact, the main practical testing scenario might involve validating the provisioning of the proposed changes.

> In certain instances, examining a single attribute of the final state could be sufficient.

Do you mean here that you could check that your Kubernetes state is `UP` and that is fine? I think rewording to make something like that clear would be a good call. 

Also, I feel like we should answer the question "Do I need this if I’m already testing my Terraform code?", right?

Reply:
>Do you mean here that you could check that your Kubernetes state is UP and that is fine? I think rewording to make something like that clear would be a good call.

Yeah, a check like this can be enough. I could go way too theoretical, and sound hazy :/ Reworded that a bit, please let me know if that helped.

>Also, I feel like we should answer the question, "Do I need this if I’m already testing my Terraform code?", right?

I tried to draw a line between infra validation and testing (cause I see that people often don't separate these terms) and to show that `check` boosts infra validation (by design) AND, in some cases, serves testing purposes.
I feel like the answer to this question is open and should be left to the reader. If we provide an answer at this point, it might look like a too-early summary.
We can rename the chapter if this question sounds confusing. Does all the above make sense to you?